api/v1alpha1/application.go

```diff
@@ -3,35 +3,68 @@ package v1alpha1
 import (
 	"golang.org/x/exp/constraints"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-//+kubebuilder:object:root=true
+// +kubebuilder:object:root=true
 type ApplicationList struct {
 	metav1.TypeMeta `json:",inline"`
 	metav1.ListMeta `json:"metadata,omitempty"`
 
 	Items []Application `json:"items"`
 }
 
-//+kubebuilder:object:root=true
-//+kubebuilder:resource:shortName="app"
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:shortName="app"
 type Application struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
 	Spec ApplicationSpec `json:"spec,omitempty"`
 }
 
-//+kubebuilder:object:generate=true
+// +kubebuilder:object:generate=true
+type ResourceRequirements struct {
+	//+kubebuilder:validation:Required
+	Limits LimitResourceList `json:"limits"`
+	//+kubebuilder:validation:Required
+	Requests RequestsResourceList `json:"requests"`
+}
+
+// +kubebuilder:object:generate=true
+type LimitResourceList struct {
+	//+kubebuilder:validation:Optional
+	CpuLimit resource.Quantity `json:"cpu,omitempty"`
+	//+kubebuilder:validation:Required
+	MemoryLimit resource.Quantity `json:"memory"`
+	//+kubebuilder:validation:Optional
+	StorageLimit resource.Quantity `json:"storage,omitempty"`
+	//+kubebuilder:validation:Optional
+	EphemeralStorageLimit resource.Quantity `json:"ephemeralStorage,omitempty"`
+}
+
+// +kubebuilder:object:generate=true
+type RequestsResourceList struct {
+	//+kubebuilder:validation:Required
+	CpuRequest resource.Quantity `json:"cpu"`
+	//+kubebuilder:validation:Required
+	MemoryRequest resource.Quantity `json:"memory"`
+	//+kubebuilder:validation:Optional
+	StorageRequest resource.Quantity `json:"storage,omitempty"`
+	//+kubebuilder:validation:Optional
+	EphemeralStorageRequest resource.Quantity `json:"ephemeralStorage,omitempty"`
+}
+
+// +kubebuilder:object:generate=true
 type ApplicationSpec struct {
 	//+kubebuilder:validation:Required
 	Image string `json:"image"`
 	//+kubebuilder:validation:Optional
 	Command []string `json:"command,omitempty"`
 
-	//+kubebuilder:validation:Optional
-	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+	//+kubebuilder:validation:Required
+	Resources *ResourceRequirements `json:"resources"`
```

Comment:
Note that this change needs to be advertised in #gen-skip when it gets merged. If anyone has defined resources today it might mean that they are unable to deploy any new code before they add that field because the validation says it's required.